klarna-on-demand/src/main/java/com/klarna/ondemand/CryptoImpl.java

```diff
@@ -0,0 +1,116 @@
+package com.klarna.ondemand;
+
+import android.content.*;
+import android.content.Context;
+import android.util.Base64;
+
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+
+class CryptoImpl implements Crypto {
+    private static final String PUBLIC_KEY = "PublicKey";
+    private static final String PRIVATE_KEY = "PrivateKey";
+    private static final String PREFERENCES_FILE_NAME = "KeyPair";
+    private static final String ALGORITHM = "RSA";
+
+    private static CryptoImpl cyptoInstance;
+
+    private String publicKeyBase64Str;
+    private PublicKey publicKey;
+    private PrivateKey privateKey;
+
+
+    public static Crypto getInstance(android.content.Context context) {
+        if (cyptoInstance == null) {
+            try {
+                cyptoInstance = new CryptoImpl(context);
+            } catch (Exception e) {
+                throw new RuntimeException("Could not initialize " + CryptoImpl.class.getName(), e);
+            }
+        }
+        return cyptoInstance;
+    }
+
+    private CryptoImpl(android.content.Context context) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        SharedPreferences sharedPreferences = context.getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE);
+
+        publicKey = readPublicKey(sharedPreferences);
+        privateKey = readPrivateKey(sharedPreferences);
+        if (publicKey == null || privateKey == null) {
+            KeyPair keyPair = generateKeyPair();
+            publicKey = keyPair.getPublic();
+            privateKey = keyPair.getPrivate();
+
+            persistKeyPair(sharedPreferences, keyPair);
+        }
+
+        publicKeyBase64Str = new String(Base64.encode(publicKey.getEncoded(), Base64.DEFAULT));
+    }
+
+    @Override
+    public String getPublicKeyBase64Str() {
+        return publicKeyBase64Str;
+    }
+
+    @Override
+    public String sign(String message) throws NoSuchAlgorithmException, InvalidKeyException, SignatureException {
```

Comment:
iOS's corresponding method's name should be changed from `getSignatureWithData` to `signWithData`